### PR TITLE
Utf 8 SubdivisionCodes

### PR DESCRIPTION
**Pull request by AliceInHunterland**
ref. https://github.com/nspcc-dev/neofs-node/issues/2637 
Can we rely on the data, which @carpawell proposed? Also, I found https://www.ip2location.com/free/iso3166-2, but it consists of less information.

> So the problem is in the "long" description? That is strange. Seems like SubdivisionCodes.csv is the only file with encoding problems. But there is some UTF-8 version of it: https://github.com/datasets/un-locode.

### CODE REVIEW

**Comment by carpawell** [comment]
> Can we rely on the data, which @carpawell proposed?

If you want to do it, double-check it, please. The only thing I wanted to say was that it is possible to format it and somebody has done it successfully. We should either find a good and _available_ source of utf-8 files or make it possible to format ourselves _automatically_ cause the database is updated at least once per year.

BTW, it is still an unknown encoding to me in the `SubdivisionCodes.csv` file.

**Comment by AliceInHunterland** [comment]
@carpawell @roman-khimov i've converted our csv to valid utf-8 strings, it rest some mismatches in symbols (less than 100 records have "?" inside) and can be overwrite manually, but it wont break IR as it does now.

File: internal/parsers/table/csv/calls.go [hunk 1]
```diff
@@ -45,8 +48,7 @@ const (
 	subDivCountry
 	subDivSubdivision
 	subDivName
-	_ // subDivLevel
-
+	_
```
Comment (roman-khimov) [inline internal/parsers/table/csv/calls.go:51]: Why this change?

File: internal/parsers/table/csv/calls.go [hunk 2]
```diff
@@ -79,16 +81,38 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
 	return rec.name, nil
 }
 
+func utf8encoding(subdivCountry string) (string, error) {
+	res, err := charmap.Windows1256.NewDecoder().String(subdivCountry)
+	if err == nil && !strings.Contains(res, "\uFFFD") && !strings.Contains(res, "\u0000") && !strings.Contains(res, "?") {
+		return res, nil
+	}
+
+	res, err = charmap.ISO8859_1.NewDecoder().String(subdivCountry)
```
Comment (roman-khimov) [inline internal/parsers/table/csv/calls.go:90]: I'd try this first and 1256 then. But maybe you're right, I don't know.

File: pkg/locodedb/calls_test.go [hunk 3]
```diff
@@ -28,9 +30,40 @@ func TestGet(t *testing.T) {
...
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
+		})
+
+	})
 	t.Run("valid key", func(t *testing.T) {
 		_, err := locodedb.NewKey("RU", "MOW")
 		require.NoError(t, err)
-
```
Comment (roman-khimov) [inline pkg/locodedb/calls_test.go:34]: Separate change.

File: pkg/locodedb/calls_test.go [hunk 4]
```diff
@@ -28,9 +30,40 @@ func TestGet(t *testing.T) {
...
+			rec, err := locodedb.Get("FRXGS")
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
+		})
+		t.Run("utf-8 locode", func(t *testing.T) {
+			rec, err := locodedb.Get("BRSPY")
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
```
Comment (roman-khimov) [inline pkg/locodedb/calls_test.go:61]: No!

File: pkg/locodedb/calls_test.go [hunk 5]
```diff
@@ -28,9 +30,40 @@ func TestGet(t *testing.T) {
...
+			fmt.Println(rec)
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
+		})
+		t.Run("utf-8 locode", func(t *testing.T) {
+			rec, err := locodedb.Get("FRXGS")
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
```
Comment (roman-khimov) [inline pkg/locodedb/calls_test.go:55]: Can't be !

File: pkg/locodedb/calls_test.go [hunk 6]
```diff
@@ -28,9 +30,40 @@ func TestGet(t *testing.T) {
...
+		})
+		t.Run("utf-8 locode", func(t *testing.T) {
+			rec, err := locodedb.Get("JOSAH")
+			fmt.Println(rec)
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
```
Comment (roman-khimov) [inline pkg/locodedb/calls_test.go:49]: This way!

File: pkg/locodedb/calls_test.go [hunk 7]
```diff
@@ -28,9 +30,40 @@ func TestGet(t *testing.T) {
...
+	t.Run("utf-8 subdiv locode", func(t *testing.T) {
+		rec, err := locodedb.Get("SESTO")
+		require.NoError(t, err)
+		require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
+		t.Run("utf-8 locode", func(t *testing.T) {
+			rec, err := locodedb.Get("ARADS")
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+
```
Comment (roman-khimov) [inline pkg/locodedb/calls_test.go:42]: Impossible!

File: internal/parsers/table/csv/calls.go [hunk 8]
```diff
@@ -79,16 +81,38 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
...
 func (t *Table) initSubDiv() (err error) {
 	t.subDivOnce.Do(func() {
 		t.mSubDiv = make(map[subDivKey]subDivRecord)
 
 		err = t.scanWords([]string{t.subDivPath}, subDivFldNum, func(words []string) error {
+			subdiv := words[subDivName]
+			if !utf8.ValidString(subdiv) {
+				subdiv, err = utf8encoding(subdiv)
+				if err != nil {
+					return err
```
Comment (roman-khimov) [inline internal/parsers/table/csv/calls.go:107]: So the process will fail? Or we're going to have some warning printed for this record? What's gonna happen?

File: internal/parsers/table/csv/calls.go (hunk 2 above)
Comment (AliceInHunterland) [reply internal/parsers/table/csv/calls.go:90]: tried this way, but charmap.ISO8859_1 leaves move unknown/ unrelevant symbols, and doesn't give any error on it..

File: internal/parsers/table/csv/calls.go (hunk 8 above)
Comment (AliceInHunterland) [reply internal/parsers/table/csv/calls.go:107]: process will fail

Comment (roman-khimov) [reply internal/parsers/table/csv/calls.go:107]: But we have to have some DB in the end, don't we?

Comment (AliceInHunterland) [reply internal/parsers/table/csv/calls.go]: changed to printing warning about the record

File: pkg/locodedb/calls_test.go [hunk 9]
```diff
@@ -28,6 +29,33 @@ func TestGet(t *testing.T) {
...
+			rec, err := locodedb.Get("FRXGS")
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+		})
+		t.Run("utf-8 locode", func(t *testing.T) {
+			rec, err := locodedb.Get("BRSPY")
+			require.NoError(t, err)
+			require.Equal(t, utf8.ValidString(rec.SubDivName), true)
+		})
+
```
Comment (roman-khimov) [inline pkg/locodedb/calls_test.go:57]: Again.

File: internal/parsers/table/csv/calls.go [hunk 10]
```diff
@@ -79,16 +83,38 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
...
 func (t *Table) initSubDiv() (err error) {
 	t.subDivOnce.Do(func() {
 		t.mSubDiv = make(map[subDivKey]subDivRecord)
 
 		err = t.scanWords([]string{t.subDivPath}, subDivFldNum, func(words []string) error {
+			subdiv := words[subDivName]
+			if !utf8.ValidString(subdiv) {
+				subdiv, err = utf8encoding(subdiv)
+				if err != nil {
+					fmt.Println(words, err)
```
Comment (roman-khimov) [inline internal/parsers/table/csv/calls.go:109]: And return nil? We still can't add it.

Comment (AliceInHunterland) [reply internal/parsers/table/csv/calls.go]: better?

File: internal/parsers/table/csv/calls.go [hunk 11]
```diff
@@ -79,19 +83,42 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
...
 		err = t.scanWords([]string{t.subDivPath}, subDivFldNum, func(words []string) error {
+			subdiv := words[subDivName]
+			if !utf8.ValidString(subdiv) {
+				subdiv, err = utf8encoding(subdiv)
+			}
+
 			t.mSubDiv[subDivKey{
 				countryCode: words[subDivCountry],
 				subDivCode:  words[subDivSubdivision],
 			}] = subDivRecord{
```
Comment (roman-khimov) [inline internal/parsers/table/csv/calls.go:113]: But you still add it here. While the name is incorrect and we know about it. Error should be handled immediately, I think.

**Comment by carpawell** [comment]
> i've converted our csv to valid utf-8 strings

@AliceInHunterland, how did you do this? Manually?

**Comment by AliceInHunterland** [comment]
> > i've converted our csv to valid utf-8 strings
> 
> @AliceInHunterland, how did you do this? Manually?

with charmap.ISO8859_1 and charmap.Windows1256

**Review by roman-khimov — DISMISSED** [review]

**Comment by carpawell** [comment]
> with charmap.ISO8859_1 and charmap.Windows1256

Oh, sorry, had not looked at the code yet when wrote the comment.

File: internal/parsers/table/csv/calls.go [hunk 12]
```diff
@@ -79,19 +83,42 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
 	return rec.name, nil
 }
 
+func utf8encoding(subdivCountry string) (string, error) {
+	res, err := charmap.Windows1256.NewDecoder().String(subdivCountry)
+	if err == nil && !strings.Contains(res, "\uFFFD") && !strings.Contains(res, "\u0000") && !strings.Contains(res, "?") {
```
Comment (carpawell) [inline internal/parsers/table/csv/calls.go:88]: can we place this duplication check to some well-named func?

File: internal/parsers/table/csv/calls.go [hunk 13]
```diff
@@ -79,19 +83,42 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
...
+			if !utf8.ValidString(subdiv) {
+				subdiv, err = utf8encoding(subdiv)
 			}
-
-			return nil
+			if err != nil {
+				if errors.Is(err, errInvalidSubName) {
+					return nil
+				}
+			} else {
```
Comment (carpawell) [inline internal/parsers/table/csv/calls.go:113]: i would improve this a little:

```go
if err != nil {
	if errors.Is(err, errInvalidSubName) {
		return nil
	}
	return err
}
t.mSubDiv[subDivKey{
	countryCode: words[subDivCountry],
	subDivCode:  words[subDivSubdivision],
}] = subDivRecord{
	name: subdiv,
}
```

**Review by carpawell — APPROVED** [review]

File: pkg/locodedb/calls.go [hunk 14]
```diff
@@ -23,7 +24,8 @@ func Get(locodeStr string) (Record, error) {
 	if err := initLocodeData(); err != nil {
 		return Record{}, err
 	}
-
+	locodeStr = strings.ReplaceAll(locodeStr, " ", "")
+	locodeStr = strings.ReplaceAll(locodeStr, "-", "")
```
Comment (roman-khimov) [inline pkg/locodedb/calls.go:28]: 1. It's a completely separate change -> different commit at least.
2. It's not entirely correct, I can pass `   XX---YYY - - - -` this way.
3. It's suboptimal, you don't need additional allocations here, just split the string properly depending on the input (that is either a 5 or 6 chars).

File: internal/parsers/table/csv/calls.go [hunk 15]
```diff
@@ -79,18 +83,45 @@ func (t *Table) SubDivName(countryCode *locodedb.CountryCode, code string) (stri
...
+				}
+			}
 			t.mSubDiv[subDivKey{
 				countryCode: words[subDivCountry],
 				subDivCode:  words[subDivSubdivision],
 			}] = subDivRecord{
-				name: words[subDivName],
+				name: subdiv,
 			}
-
```
Comment (roman-khimov) [inline internal/parsers/table/csv/calls.go:93]: Whitespace change.

File: internal/parsers/table/csv/calls.go (hunk 13 above)
Comment (AliceInHunterland) [reply internal/parsers/table/csv/calls.go]: updated

File: pkg/locodedb/calls.go (hunk 14 above)
Comment (AliceInHunterland) [reply pkg/locodedb/calls.go]: opened different pr

**Review by roman-khimov — APPROVED** [review]
